Make dot and vdot structured ops (#64)

matthewhagraphcore opened the pull request:
This PR refactors the dot and vdot operations in PyTorch to use a structured kernel approach for both CPU and CUDA implementations. The changes involve converting the existing kernels into structured kernels. Both the CPU and CUDA kernels have been updated to use the `TORCH_IMPL_FUNC` and `TORCH_META_FUNC` macros, ensuring they follow the structured kernel pattern.

I moved the meat of the kernels from `aten/src/ATen/native/Blas.cpp` to `aten/src/ATen/native/LinearAlgebra.cpp`, keeping as much of the original kernel as possible, however a bit had to be adapted as they now became `out` variants.

I have provided some CPU bechmarks below to ensure that the minor changes did not impact performance. 

Testing:
The changes are covered by existing tests for the dot operation.

matthewhagraphcore committed:
PYT-194 make dot a structured op (#64)

This PR refactors the dot operation in PyTorch to use a structured
kernel approach for both CPU and CUDA implementations. The changes
involve converting the existing kernels into structured kernels. Both
the CPU and CUDA kernels have been updated to use the `TORCH_IMPL_FUNC`
and `TORCH_META_FUNC` macros, ensuring they follow the structured kernel
pattern.

Jira-task: PYT-194

Testing:
The changes are covered by existing tests for the dot operation.

matthewhagraphcore committed:
PYT-194 make dot a structured op (#64)

This PR refactors the dot operation in PyTorch to use a structured
kernel approach for both CPU and CUDA implementations. The changes
involve converting the existing kernels into structured kernels. Both
the CPU and CUDA kernels have been updated to use the `TORCH_IMPL_FUNC`
and `TORCH_META_FUNC` macros, ensuring they follow the structured kernel
pattern.

Jira-task: PYT-194

Testing:
The changes are covered by existing tests for the dot operation.

matthewhagraphcore commented:
Pretty sure the `native_functions.yaml` FC stuff doesn't apply here.

syed-ahmed commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -656,11 +656,14 @@ inline void dot_check(const Tensor& self, const Tensor& other) {
-        return (dot_cuda(self.conj(), other.conj())).conj();
+        auto temp_result = at::empty({}, self.options());
+        at::dot_out(temp_result, self.conj(), other.conj());
+        result.fill_(temp_result.conj());
```
Does this change the performance characteristics of this kernel? Could you please provide some profiling results?

syed-ahmed replied:
You can follow the instructions here: https://github.com/pytorch/pytorch/blob/main/CONTRIBUTING.md#cuda-development-tips (I already see below you have done something similar for the CPU path).

AnthonyBarbier replied:
> If you are working on the CUDA code, 
Except we're not working on CUDA code here and we don't have access to build machines with CUDA HW.

TBH it doesn't matter if we did: if this part of the code is indeed performance critical then you must have a test suite and a workflow that regularly checks for regressions, right? (That's why I was asking which workflow we need to trigger against this PR)
If you don't it's probably easier to do like for other PRs: merge it and revert if it causes issues in the nightlies?

syed-ahmed commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1899,20 +1929,17 @@ TORCH_IMPL_FUNC(bmm_out_cpu)
-  TORCH_CHECK(
-    (output_device == input1_device) && (input1_device == input2_device),
-    "dot: Expected the output and input tensors to be on the "
-    "same device, but got the output tensor on ", output_device,
```
Seems like this device check is lost? Or is it captured somewhere else?

charlie-wt replied:
from what i've seen from the generated code in `build/aten/src/ATen/Register*.cpp`, making an op structured doesn't add any automatic device checks, it'll just add the scaffolding to be able to write a `TORCH_META_FUNC` & do what you want in it (which may include a device check).

the automatic device check generation seems to be controlled by whether you have `device_check: NoCheck` in the entry in native_functions.yaml, which `dot` & `vdot` don't have so i'd expect they're already generating device checks. those checks aren't enabled for all backends, though; according to [this comment](https://github.com/pytorch/pytorch/blob/50efbb9f1e7111b4b6d5b8e9a6064ee9783930be/torchgen/model.py#L1261-L1264), CUDA & HIP are the only in-tree backends that'll generate them automatically, which i guess is a divergence from the current behaviour of checking regardless of dispatched backend.

i could be wrong on this though, that's all just based on poking around a bit in the generated code.

matthewhagraphcore replied:
OK, I will add the device check back in

syed-ahmed commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -680,16 +688,15 @@ Tensor dot_cuda(const Tensor& self, const Tensor& other) {
-}
+  if (self._is_zerotensor() || other._is_zerotensor()) {
+    result.fill_(0);
+    return;
```
Same comment as before: how does this change the performance of this kernel?

AnthonyBarbier replied:
Same as above: could you please point us at which workflow we need to trigger to get those numbers?

syed-ahmed replied:
https://github.com/pytorch/pytorch/blob/main/CONTRIBUTING.md#cuda-development-tips

syed-ahmed replied:
The behavior is changed - before, this path would use ZeroTensor allocator: https://github.com/pytorch/pytorch/blob/ff616c26fb66b26966b6f6f50aa85bf615400816/aten/src/ATen/native/TensorFactories.h#L119-L120, which seems to not fill result with 0.

syed-ahmed commented on torch/csrc/jit/runtime/static/generated_ops.cpp:
```diff
@@ -783,12 +783,12 @@ REGISTER_OPERATOR_FUNCTOR(aten::dot, aten_dot, [](Node* n) -> SROperator {
       const auto& tensor = p_node->Input(1).toTensor();
       if (p_node->Output(0).isNone()) {
-        p_node->Output(0) = at::native::dot(self, tensor);
+        p_node->Output(0) = at::cpu::dot(self, tensor);
```
I'm curious, how did the namespace change?

matthewhagraphcore replied:
That's just the namespace that the structured ops generate the kernels into. Compare with erf below which is also structured. The `at::native::dot` function is no longer required as the device guard and other functionality has moved to `META_FUNC`

syed-ahmed commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1899,20 +1929,17 @@ TORCH_IMPL_FUNC(bmm_out_cpu)
+  // Perform the matrix multiplication
+  auto self_reshaped = self.view({1, -1});  // Reshape self to be a 1xN matrix
+  auto other_reshaped = other.view({-1, 1});  // Reshape other to be a Nx1 matrix
+  auto matmul_result = at::matmul(self_reshaped, other_reshaped);
```
I'm curious, why is this change needed? Also how does `self.dot(other)` compare to view+view+matmul in terms of performance?

matthewhagraphcore replied:
I've adapted this to match the original impl

syed-ahmed reviewed (changes requested):
I have some questions about performance. It would be great if you could provide some profiling results. I haven't thoroughly checked the structured kernel approach.

matthewhagraphcore committed:
Fix paths in gen_static_runtime_ops

matthewhagraphcore committed:
Fix paths in gen_static_runtime_ops

matthewhagraphcore committed:
First pass at vdot structured op

matthewhagraphcore committed:
First pass at vdot structured op

matthewhagraphcore committed:
corrections

matthewhagraphcore committed:
corrections

matthewhagraphcore committed:
comment and remove unused code

matthewhagraphcore committed:
comment and remove unused code

matthewhagraphcore committed:
Merge remote-tracking branch 'origin/main' into dot_a_structured_op

matthewhagraphcore committed:
Use kernels from BLAS

matthewhagraphcore committed:
Use kernels from BLAS

matthewhagraphcore commented:
CPU benchmarking:

script: 
```
import torch
import torch.utils.benchmark as benchmark

device = "cpu"

results = []

# Define sizes for the vectors to benchmark
vector_sizes = [0, 1024, 65536, 1048576]

for vector_size in vector_sizes:
    print(f"vector_size: {vector_size}")

    # Generate a pair of vectors using arange
    vec1 = torch.arange(vector_size, dtype=torch.float32, device=device)
    vec2 = torch.arange(vector_size, dtype=torch.float32, device=device)

    t = benchmark.Timer(
            stmt='torch.dot(vec1, vec2)',
            label='torch.dot, arange vectors',
            globals={"vec1": vec1, "vec2": vec2},
            description=f"Size {vector_size}",
        ).blocked_autorange(min_run_time=5)
    results.append(t)

    # Generate a pair of zero tensors
    vec1_zero = torch.zeros(vector_size, dtype=torch.float32, device=device)
    vec2_zero = torch.zeros(vector_size, dtype=torch.float32, device=device)

    t_zero = benchmark.Timer(
            stmt='torch.dot(vec1_zero, vec2_zero)',
            label='torch.dot, zero vectors',
            globals={"vec1_zero": vec1_zero, "vec2_zero": vec2_zero},
            description=f"Size {vector_size}",
        ).blocked_autorange(min_run_time=5)
    results.append(t_zero)

    # Generate a pair of complex tensors
    vec1_complex = torch.rand(vector_size, dtype=torch.complex64, device=device)
    vec2_complex = torch.rand(vector_size, dtype=torch.complex64, device=device)

    t_complex = benchmark.Timer(
            stmt='torch.dot(vec1_complex, vec2_complex)',
            label='torch.dot, complex vectors',
            globals={"vec1_complex": vec1_complex, "vec2_complex": vec2_complex},
            description=f"Size {vector_size}",
        ).blocked_autorange(min_run_time=5)
    results.append(t_complex)

compare = benchmark.Compare(results)
compare.trim_significant_figures()
compare.colorize(rowwise=True)
compare.print()

```

main:
```

[--------------------------- torch.dot, arange vectors ----------------------------]
                             |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -------------------------------------------------------------------------
      torch.dot(vec1, vec2)  |    19    |      20     |     31.1     |      191     

Times are in microseconds (us).

[--------------------------------- torch.dot, zero vectors ----------------------------------]
                                       |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -----------------------------------------------------------------------------------
      torch.dot(vec1_zero, vec2_zero)  |    19    |      20     |     30.9     |      190     

Times are in microseconds (us).

[----------------------------------- torch.dot, complex vectors -----------------------------------]
                                             |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -----------------------------------------------------------------------------------------
      torch.dot(vec1_complex, vec2_complex)  |    19    |      20     |     45.3     |      1000    

```
this PR:
```

[--------------------------- torch.dot, arange vectors ----------------------------]
                             |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -------------------------------------------------------------------------
      torch.dot(vec1, vec2)  |    18    |      17     |     27.7     |      180     

Times are in microseconds (us).

[--------------------------------- torch.dot, zero vectors ----------------------------------]
                                       |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -----------------------------------------------------------------------------------
      torch.dot(vec1_zero, vec2_zero)  |    16    |      17     |     28.2     |      186     

Times are in microseconds (us).

[----------------------------------- torch.dot, complex vectors -----------------------------------]
                                             |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -----------------------------------------------------------------------------------------
      torch.dot(vec1_complex, vec2_complex)  |    17    |      17     |     42.3     |      999     

```

matthewhagraphcore commented on torchgen/static_runtime/gen_static_runtime_ops.py:
```diff
@@ -137,19 +137,19 @@ def main() -> None:
-        default="caffe2/aten/src/ATen",
+        default="aten/src/ATen",
     )
     parser.add_argument(
         "-p",
         "--generated-ops-cpp-path",
         help="path to directory to generate op dispatcher .cpp file",
-        default="caffe2/torch/csrc/jit/runtime/static/generated_ops.cpp",
+        default="torch/csrc/jit/runtime/static/generated_ops.cpp",
     )
     parser.add_argument(
         "-t",
         "--generated-ops-test-cpp-path",
         help="path to directory to generate op dispatcher .cpp file",
-        default="caffe2/benchmarks/static_runtime/test_generated_ops.cc",
+        default="benchmarks/static_runtime/test_generated_ops.cc",
```
These paths have been incorrect for some time by the looks of things, so have taken the opportunity to correct them.

matthewhagraphcore committed:
Merge remote-tracking branch 'origin/main' into dot_a_structured_op

matthewhagraphcore committed:
Merge remote-tracking branch 'origin/main' into dot_a_structured_op

matthewhagraphcore committed:
Re-add device checks

matthewhagraphcore committed:
Re-add device checks

matthewhagraphcore committed:
Merge branch 'pytorch:main' into dot_a_structured_op

matthewhagraphcore commented:
@janeyx99 could you get someone to review this? Thanks

janeyx99 commented:
it looks like @syed-ahmed @charlie-wt and @AnthonyBarbier are already on it!

AnthonyBarbier reviewed (approved):
Looks good to me

huydhn committed:
Fix test_skip_data_serialization pickle exception match (#136617)

The test is failing in trunk atm with the following error:

```
test_serialization.py::TestSerialization::test_skip_data_serialization_materialize_fake_False - AssertionError: "Can't pickle local object 'WeakValueDictionary.__init__.<locals>.remove'" does not match "Can't get local object 'WeakValueDictionary.__init__.<locals>.remove'"
```

for example, https://hud.pytorch.org/pytorch/pytorch/commit/36f0e6116667c5391405e9be194b5aae3688f6dc

This comes from this cpython commit https://github.com/python/cpython/commit/a3076c734d310d2b860c6da5e10a67fc9c29d9b6, and manifests in python 3.12.5 currently used in CI.  The failure doesn't happen when I try it out with 3.12.3 and 3.12.4.  Looking at the commit logs of https://github.com/python/cpython/commits/main/Lib/pickle.py, it looks like the exception message is changing back and forth, so I guess a regex match would capture both.

syed-ahmed commented:
> Except we're not working on CUDA code here

@AnthonyBarbier May be I'm missing something here. I see changes in `aten/src/ATen/native/cuda/Blas.cpp`. That's why I asked to see some sort of benchmarking before and after, like it's done for the CPU side. May be you could add a test temporarily in test/test_cuda.py and report the numbers you are seeing on the github CI run which has the CUDA hw? (feel free to ping here and we can trigger the workflow).

matthewhagraphcore committed:
Merge branch 'main' into dot_a_structured_op

matthewhagraphcore committed:
Merge branch 'main' into dot_a_structured_op

matthewhagraphcore committed:
Merge remote-tracking branch 'myfork/dot_a_structured_op' into dot_a_structured_op

Commit:
Merge branch 'main' into dot_a_structured_op

matthewhagraphcore committed:
Correct vdot_cuda

matthewhagraphcore committed:
Correct vdot_cuda

matthewhagraphcore commented:
I'm not sure what the easyCLA issue is, I have used the same email throughout? 

@syed-ahmed I have finally got round to doing the benchmarks as follows: 

<details><summary>dot CUDA benchmarks</summary>
<p>

```
import torch
import torch.utils.benchmark as benchmark

device = "cuda"

results = []

# Define sizes for the vectors to benchmark
vector_sizes = [0, 1024, 65536, 1048576]

for vector_size in vector_sizes:
    print(f"vector_size: {vector_size}")

    # Generate a pair of vectors using arange
    vec1 = torch.arange(vector_size, dtype=torch.float32, device=device)
    vec2 = torch.arange(vector_size, dtype=torch.float32, device=device)

    t = benchmark.Timer(
            stmt='torch.dot(vec1, vec2)',
            label='torch.dot, arange vectors',
            globals={"vec1": vec1, "vec2": vec2},
            description=f"Size {vector_size}",
        ).blocked_autorange(min_run_time=5)
    results.append(t)

    # Generate a pair of zero tensors
    vec1_zero = torch.zeros(vector_size, dtype=torch.float32, device=device)
    vec2_zero = torch.zeros(vector_size, dtype=torch.float32, device=device)

    t_zero = benchmark.Timer(
            stmt='torch.dot(vec1_zero, vec2_zero)',
            label='torch.dot, zero vectors',
            globals={"vec1_zero": vec1_zero, "vec2_zero": vec2_zero},
            description=f"Size {vector_size}",
        ).blocked_autorange(min_run_time=5)
    results.append(t_zero)

    # Generate a pair of complex tensors
    vec1_complex = torch.rand(vector_size, dtype=torch.complex64, device=device)
    vec2_complex = torch.rand(vector_size, dtype=torch.complex64, device=device)

    t_complex = benchmark.Timer(
            stmt='torch.dot(vec1_complex, vec2_complex)',
            label='torch.dot, complex vectors',
            globals={"vec1_complex": vec1_complex, "vec2_complex": vec2_complex},
            description=f"Size {vector_size}",
        ).blocked_autorange(min_run_time=5)
    results.append(t_complex)

compare = benchmark.Compare(results)
compare.trim_significant_figures()
compare.colorize(rowwise=True)
compare.print()
```
main:
```
[--------------------------- torch.dot, arange vectors ----------------------------]
                             |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -------------------------------------------------------------------------
      torch.dot(vec1, vec2)  |    17    |      23     |      12      |      35.7    

Times are in microseconds (us).

[--------------------------------- torch.dot, zero vectors ----------------------------------]
                                       |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -----------------------------------------------------------------------------------
      torch.dot(vec1_zero, vec2_zero)  |   6.8    |      12     |      12      |      35.8    

Times are in microseconds (us).

[----------------------------------- torch.dot, complex vectors -----------------------------------]
                                             |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -----------------------------------------------------------------------------------------
      torch.dot(vec1_complex, vec2_complex)  |   6.9    |      12     |      12      |      65.0    

Times are in microseconds (us).
```
this pr:
```
[--------------------------- torch.dot, arange vectors ----------------------------]
                             |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -------------------------------------------------------------------------
      torch.dot(vec1, vec2)  |   6.5    |      11     |      11      |      35.7    

Times are in microseconds (us).

[--------------------------------- torch.dot, zero vectors ----------------------------------]
                                       |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -----------------------------------------------------------------------------------
      torch.dot(vec1_zero, vec2_zero)  |   6.6    |      11     |      12      |      35.8    

Times are in microseconds (us).

[----------------------------------- torch.dot, complex vectors -----------------------------------]
                                             |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -----------------------------------------------------------------------------------------
      torch.dot(vec1_complex, vec2_complex)  |    7     |      11     |      11      |      65.0    
```
</p>
</details> 


<details><summary>vdot CUDA benchmarks</summary>
<p>

```
import torch
import torch.utils.benchmark as benchmark

device = "cuda"

results = []

# Define sizes for the vectors to benchmark
vector_sizes = [0, 1024, 65536, 1048576]

for vector_size in vector_sizes:
    print(f"vector_size: {vector_size}")

    # Generate a pair of vectors using arange
    vec1 = torch.arange(vector_size, dtype=torch.float32, device=device)
    vec2 = torch.arange(vector_size, dtype=torch.float32, device=device)

    t = benchmark.Timer(
            stmt='torch.vdot(vec1, vec2)',
            label='torch.vdot, arange vectors',
            globals={"vec1": vec1, "vec2": vec2},
            description=f"Size {vector_size}",
        ).blocked_autorange(min_run_time=5)
    results.append(t)

    # Generate a pair of zero tensors
    vec1_zero = torch.zeros(vector_size, dtype=torch.float32, device=device)
    vec2_zero = torch.zeros(vector_size, dtype=torch.float32, device=device)

    t_zero = benchmark.Timer(
            stmt='torch.vdot(vec1_zero, vec2_zero)',
            label='torch.vdot, zero vectors',
            globals={"vec1_zero": vec1_zero, "vec2_zero": vec2_zero},
            description=f"Size {vector_size}",
        ).blocked_autorange(min_run_time=5)
    results.append(t_zero)

    # Generate a pair of complex tensors
    vec1_complex = torch.rand(vector_size, dtype=torch.complex64, device=device)
    vec2_complex = torch.rand(vector_size, dtype=torch.complex64, device=device)

    t_complex = benchmark.Timer(
            stmt='torch.vdot(vec1_complex, vec2_complex)',
            label='torch.vdot, complex vectors',
            globals={"vec1_complex": vec1_complex, "vec2_complex": vec2_complex},
            description=f"Size {vector_size}",
        ).blocked_autorange(min_run_time=5)
    results.append(t_complex)

compare = benchmark.Compare(results)
compare.trim_significant_figures()
compare.colorize(rowwise=True)
compare.print()
```
main:
```
[---------------------------- torch.vdot, arange vectors ---------------------------]
                              |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: --------------------------------------------------------------------------
      torch.vdot(vec1, vec2)  |   6.9    |      12     |      12      |      35.7    

Times are in microseconds (us).

[---------------------------------- torch.vdot, zero vectors ---------------------------------]
                                        |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: ------------------------------------------------------------------------------------
      torch.vdot(vec1_zero, vec2_zero)  |   6.9    |      12     |      12      |      35.8    

Times are in microseconds (us).

[----------------------------------- torch.vdot, complex vectors -----------------------------------]
                                              |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: ------------------------------------------------------------------------------------------
      torch.vdot(vec1_complex, vec2_complex)  |   6.8    |      12     |      12      |      65.0    

Times are in microseconds (us).
```
this pr:
```
[---------------------------- torch.vdot, arange vectors ---------------------------]
                              |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: --------------------------------------------------------------------------
      torch.vdot(vec1, vec2)  |    11    |      11     |      16      |      35.7    

Times are in microseconds (us).

[---------------------------------- torch.vdot, zero vectors ---------------------------------]
                                        |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: ------------------------------------------------------------------------------------
      torch.vdot(vec1_zero, vec2_zero)  |    10    |      16     |      16      |      35.7    

Times are in microseconds (us).

[----------------------------------- torch.vdot, complex vectors -----------------------------------]
                                              |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: ------------------------------------------------------------------------------------------
      torch.vdot(vec1_complex, vec2_complex)  |   6.6    |      16     |      16      |      65.0    

Times are in microseconds (us).
```


</p>
</details>

syed-ahmed commented:
For vdot it looks like it takes longer than before, for sizes < 1048576?

syed-ahmed commented:
For the easyCLA situation, may be you need to do a rebase instead of merge. So cleanup/squash some of the "merge" commits (for instance using `git rebase -i HEAD~10` and the squash/delete as necessary). And then next time you need to pull in latest changes from upstream pytorch, I usually do `git fetch origin; git rebase origin/viable/strict` on my PRs.

matthewhagraphcore commented:
> For vdot it looks like it takes longer than before, for sizes < 1048576?

Damn, yes I didn't realize that. I can't see anything obvious as to why. I will try and diagnose. Can you see anything obvious?

matthewhagraphcore committed:
Merge branch 'main' into dot_a_structured_op

syed-ahmed commented:
@matthewhagraphcore I don't see anything obvious. You'll have to potentially compare a profile (either PyTorch Profiler or Nsight profiler) before and after to find the root cause.

xw285cornell committed:
[AMD] Fix torch ck backend build with 6.2.1 (#138434)

Summary: It's complaining about missing __hip_bfloat162 definition w/o this header.

Differential Revision: D64673284

Pull Request resolved: https://github.com/pytorch/pytorch/pull/138434
Approved by: https://github.com/yaoyj11, https://github.com/houseroad

cyyever committed:
Remove torch/csrc/api/include/torch/linalg.h (#138435)

Only one place in OSS uses it.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/138435
Approved by: https://github.com/r-barnes

matthewhagraphcore committed:
Merge branch 'main' into dot_a_structured_op

matthewhagraphcore committed:
Merge branch 'dot_a_structured_op' of https://github.com/matthewhagraphcore/pytorch into dot_a_structured_op

matthewhagraphcore commented:
Hi @syed-ahmed  , I have rerun the benchmarking and seem to be getting very similar results. Perhaps my previous run was noise? 

## Benchmarking Results

### `main` Branch

<details>
<summary>Click to expand results for the `main` branch</summary>
[--------------------------- torch.dot, arange vectors ----------------------------]
                             |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -------------------------------------------------------------------------
      torch.dot(vec1, vec2)  |   7.0    |      12     |      12      |     35.7    

Times are in microseconds (us).

[---------------------------- torch.vdot, arange vectors ---------------------------]
                              |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: --------------------------------------------------------------------------
      torch.vdot(vec1, vec2)  |   7.0    |      12     |      12      |     35.7    

Times are in microseconds (us).

[--------------------------------- torch.dot, zero vectors ----------------------------------]
                                       |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -----------------------------------------------------------------------------------
      torch.dot(vec1_zero, vec2_zero)  |   6.9    |      12     |      12      |     35.7    

Times are in microseconds (us).

[---------------------------------- torch.vdot, zero vectors ---------------------------------]
                                        |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: ------------------------------------------------------------------------------------
      torch.vdot(vec1_zero, vec2_zero)  |   7.0    |      12     |      12      |     35.7    

Times are in microseconds (us).

[----------------------------------- torch.vdot, complex vectors -----------------------------------]
                                              |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: ------------------------------------------------------------------------------------------
      torch.vdot(vec1_complex, vec2_complex)  |   6.8    |      11     |      11      |     65.2    

Times are in microseconds (us).
</details>

### `pr` Branch

<details>
<summary>Click to expand results for the `pr` branch</summary>
[--------------------------- torch.dot, arange vectors ----------------------------]
                             |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -------------------------------------------------------------------------
      torch.dot(vec1, vec2)  |   6.3   |      11     |      11      |     35.7   

Times are in microseconds (us).

[---------------------------- torch.vdot, arange vectors ---------------------------]
                              |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: --------------------------------------------------------------------------
      torch.vdot(vec1, vec2)  |   6.4   |      11     |      11      |     35.8   

Times are in microseconds (us).

[--------------------------------- torch.dot, zero vectors ----------------------------------]
                                       |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: -----------------------------------------------------------------------------------
      torch.dot(vec1_zero, vec2_zero)  |   6.3   |      11     |      11      |     35.8   

Times are in microseconds (us).

[---------------------------------- torch.vdot, zero vectors ---------------------------------]
                                        |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: ------------------------------------------------------------------------------------
      torch.vdot(vec1_zero, vec2_zero)  |   6.4   |      11     |      11      |     35.8   

Times are in microseconds (us).

[----------------------------------- torch.vdot, complex vectors -----------------------------------]
                                              |  Size 0  |  Size 1024  |  Size 65536  |  Size 1048576
1 threads: ------------------------------------------------------------------------------------------
      torch.vdot(vec1_complex, vec2_complex)  |   6.4   |      11     |      11      |     65.3   

Times are in microseconds (us).
</details>

## Script: 


<details>
<summary>Click to expand for the script</summary>
import torch
import torch.utils.benchmark as benchmark
import subprocess
import sys 
import io

device = "cuda"

results = []

# Define sizes for the vectors to benchmark
vector_sizes = [0, 1024, 65536, 1048576]

# Get the current Git branch
def get_git_branch():
    try:
        # Use subprocess to get the current git branch
        result = subprocess.run(["git", "rev-parse", "--abbrev-ref", "HEAD"], stdout=subprocess.PIPE, check=True)
        return result.stdout.decode('utf-8').strip()
    except subprocess.CalledProcessError:
        return "unknown_branch"

branch = get_git_branch()

for vector_size in vector_sizes:
    print(f"vector_size: {vector_size}")

    # Generate a pair of vectors using arange
    vec1 = torch.arange(vector_size, dtype=torch.float32, device=device)
    vec2 = torch.arange(vector_size, dtype=torch.float32, device=device)

    # Benchmark torch.dot
    t_dot = benchmark.Timer(
        stmt='torch.dot(vec1, vec2)',
        label='torch.dot, arange vectors',
        globals={"vec1": vec1, "vec2": vec2},
        description=f"Size {vector_size}",
    ).blocked_autorange(min_run_time=5)
    results.append(t_dot)

    # Benchmark torch.vdot
    t_vdot = benchmark.Timer(
        stmt='torch.vdot(vec1, vec2)',
        label='torch.vdot, arange vectors',
        globals={"vec1": vec1, "vec2": vec2},
        description=f"Size {vector_size}",
    ).blocked_autorange(min_run_time=5)
    results.append(t_vdot)

    # Generate a pair of zero tensors
    vec1_zero = torch.zeros(vector_size, dtype=torch.float32, device=device)
    vec2_zero = torch.zeros(vector_size, dtype=torch.float32, device=device)

    # Benchmark torch.dot on zero vectors
    t_dot_zero = benchmark.Timer(
        stmt='torch.dot(vec1_zero, vec2_zero)',
        label='torch.dot, zero vectors',
        globals={"vec1_zero": vec1_zero, "vec2_zero": vec2_zero},
        description=f"Size {vector_size}",
    ).blocked_autorange(min_run_time=5)
    results.append(t_dot_zero)

    # Benchmark torch.vdot on zero vectors
    t_vdot_zero = benchmark.Timer(
        stmt='torch.vdot(vec1_zero, vec2_zero)',
        label='torch.vdot, zero vectors',
        globals={"vec1_zero": vec1_zero, "vec2_zero": vec2_zero},
        description=f"Size {vector_size}",
    ).blocked_autorange(min_run_time=5)
    results.append(t_vdot_zero)

    # Generate a pair of complex tensors
    vec1_complex = torch.rand(vector_size, dtype=torch.complex64, device=device)
    vec2_complex = torch.rand(vector_size, dtype=torch.complex64, device=device)

    # Benchmark torch.dot on complex vectors (will raise an error, so we'll skip this)
    # torch.dot does not support complex tensors, hence only benchmarking torch.vdot

    # Benchmark torch.vdot on complex vectors
    t_vdot_complex = benchmark.Timer(
        stmt='torch.vdot(vec1_complex, vec2_complex)',
        label='torch.vdot, complex vectors',
        globals={"vec1_complex": vec1_complex, "vec2_complex": vec2_complex},
        description=f"Size {vector_size}",
    ).blocked_autorange(min_run_time=5)
    results.append(t_vdot_complex)

# Compare the results and print them in the console
compare = benchmark.Compare(results)
compare.trim_significant_figures()
compare.colorize(rowwise=True)
compare.print()
# Save the results to a file
output_file = f"benchmark_{branch}.txt"

# Redirect the output of compare.print() to both the console and a file
with open(output_file, "w") as f:
    # Capture the printed comparison output
    original_stdout = sys.stdout
    sys.stdout = io.StringIO()  # Redirect stdout to a string buffer
    compare.print()             # Print the comparison (goes to the string buffer)
    output = sys.stdout.getvalue()  # Get the string from the buffer
    sys.stdout = original_stdout    # Restore the original stdout

    # Print the comparison to the console
    print(output)

    # Write the comparison to the file
    f.write(output)

print(f"Benchmark results saved to {output_file}")

</details>

matthewhagraphcore committed:
Merge branch 'main' into dot_a_structured_op

matthewhagraphcore commented:
Hi @syed-ahmed , could you re-review?

syed-ahmed commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1892,36 +1935,146 @@ TORCH_IMPL_FUNC(bmm_out_cpu)
+
+    if (use_mkldnn_matmul(self, other, /*result=*/Tensor())) {
+    // mkldnn matmul expect result have sizes info to create ideep tensor
+    auto r = at::empty({1, 1}, self.options());
```
nit: see note about consistent spaces/style.

syed-ahmed commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1892,36 +1935,146 @@ TORCH_IMPL_FUNC(bmm_out_cpu)
+    }
+
+    if (self._is_zerotensor() || other._is_zerotensor()) {
+    result.fill_(0);
```
nit: see note about consistent spaces/style.

syed-ahmed commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1892,36 +1935,146 @@ TORCH_IMPL_FUNC(bmm_out_cpu)
+        input2_device);
+    // Dispatch to `dot` for real dtypes.
+    if (!self.is_complex()) {
+    dot_out_impl(self, other, result);
```
nit: see note about consistent spaces/style.

syed-ahmed commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1892,36 +1935,146 @@ TORCH_IMPL_FUNC(bmm_out_cpu)
+    return;
+    }
+
+    if (self.is_conj()) {
```
nit: see note about consistent spaces/style.

syed-ahmed commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1892,36 +1935,146 @@ TORCH_IMPL_FUNC(bmm_out_cpu)
+    }
+    return;
+    } else if (other.is_conj()) {
+    Tensor temp_result = at::empty({}, result.options());
```
nit: see note about consistent spaces/style.

syed-ahmed commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1892,36 +1935,146 @@ TORCH_IMPL_FUNC(bmm_out_cpu)
+      if (other.is_conj()) {
+        Tensor temp_result = at::empty({}, result.options());
+        dot_out_impl(self.conj(), other.conj(), temp_result);
+        result.copy_(temp_result.conj());
```
I'm curious - why is it a `copy_` here and not a `fill_` (like used in some other places)?

syed-ahmed commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1892,36 +1935,146 @@ TORCH_IMPL_FUNC(bmm_out_cpu)
+    } else if (other.is_conj()) {
+    Tensor temp_result = at::empty({}, result.options());
+    dot_out_impl(self, other.conj(), result);
+    result.copy_(temp_result.conj());
```
I'm curious - why is it a `copy_` here and not a `fill_` (like used in some other places)?

matthewhagraphcore replied:
No reason. I didn't know there was a Tensor overload for fill_

syed-ahmed commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1892,36 +1935,146 @@ TORCH_IMPL_FUNC(bmm_out_cpu)
+    }
+
+    if (self._is_zerotensor() || other._is_zerotensor()) {
+    result.fill_(0);
```
The behavior is changed - before, this path would use ZeroTensor allocator: https://github.com/pytorch/pytorch/blob/ff616c26fb66b26966b6f6f50aa85bf615400816/aten/src/ATen/native/TensorFactories.h#L119-L120, which seems to not fill result with 0.

syed-ahmed commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -617,57 +617,31 @@ TORCH_IMPL_FUNC(bmm_out_cuda)(const Tensor& batch1, const Tensor& batch2, const
-       } else {
-         return vdot_cuda(self.conj(), other);
-       }
+        dot_out_cuda_impl(self.conj(), other.conj(), result);
```
Is the `.conj()` on the result lost here?

matthewhagraphcore replied:
I think I've resolved this

syed-ahmed commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -697,31 +670,43 @@ return AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(
 
   if (self._is_zerotensor() || other._is_zerotensor()) {
-    return at::_efficientzerotensor({}, self.options());
+    result.fill_(0);
```
The behavior is changed - before, this path would use ZeroTensor allocator: https://github.com/pytorch/pytorch/blob/ff616c26fb66b26966b6f6f50aa85bf615400816/aten/src/ATen/native/TensorFactories.h#L119-L120, which seems to not fill result with 0.

syed-ahmed reviewed (commented):
Thanks for the follow up @matthewhagraphcore. I did a more line-by-line review and left some comments. Could you please remind me what was the configuration of the machine the cuda benchmark was done in?

matthewhagraphcore committed:
Merge branch 'main' into dot_a_structured_op

matthewhagraphcore committed:
PR changes

matthewhagraphcore commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1892,36 +1935,146 @@ TORCH_IMPL_FUNC(bmm_out_cpu)
-    ", the 'input' tensor on ", input1_device, ", and the 'other' tensor on ", input2_device);
-  at::native::resize_output(result, {});
-  TORCH_CHECK(result.scalar_type() == self.scalar_type(),
-           "result dtype ", result.scalar_type(), " does not match input dtype ", self.scalar_type());
```
I couldn't do this in the META_FUNC as it doesn't take the result tensor, so I;ve had to awkwardly put it in the TORCH_IMPL_FUNC for each. Can you see a better way?

matthewhagraphcore commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1892,36 +1935,146 @@ TORCH_IMPL_FUNC(bmm_out_cpu)
+        dot_out_impl(self.conj(), other.conj(), temp_result);
+        result.copy_(temp_result.conj());
+      } else {
+        vdot_out_impl(other.conj(), self, result);
```
I've fixed this.

syed-ahmed commented:
Looks like the zero tensor path was added by @anjali411 in: https://github.com/pytorch/pytorch/pull/71129. @anjali411 whenever you get a chance, could you please comment if the zero tensor path is still needed?

matthewhagraphcore commented:
I'm going to close this as it's proving too difficult to test CUDA stuff , it doesn't provide too much benefit anyway.
